---
title: "Class11: Structural Bioinformatics"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## The PDB Database

The [PDB](http://www.rcsb.org/) is the main repository for biomoleculer structure data. 

Here we examine the contents of the PDB:

>Q1: Download a CSV file from the PDB site (accessible from “Analyze” -> “PDB Statistics” >
“by Experimental Method and Molecular Type”. Move this CSV file into your RStudio project
and determine the percentage of structures solved by X-Ray and Electron Microscopy. From
the website what proportion of structures are protein? Aim to have a rendered GitHub document
with working code that yields your answers

```{r}
db <- read.csv("Data Export Summary.csv", row.names=1)
head (db)

```

How many are Xray, etc...?
```{r}
(db$Total/sum(db$Total)) * 100

```

What percent are Protein...
```{r}
(sum(db$Proteins)/sum(db$Total)) * 100
```
We could also try the datapasta package and copy from website and go to "Addins > "Pasta as data.frame"...

The name Protein.NAComplex had to be adjusted. 
```{r}
library(datapasta)

tmp <- data.frame(stringsAsFactors=FALSE,
   Experimental.Method = c("X-Ray", "NMR", "Electron Microscopy", "Other",
                           "Multi Method", "Total"),
              Proteins = c(126880, 11062, 2277, 256, 129, 140604),
         Nucleic.Acids = c(2012, 1279, 31, 4, 5, 3331),
    ProteinComplex = c(6547, 259, 800, 6, 2, 7614),
                 Other = c(8, 8, 0, 13, 1, 30),
                 Total = c(135447, 12608, 3108, 279, 137, 151579)
)

```

>Q2: Type HIV in the PDB website search box on the home page and determine how many HIV-1 protease structures are in the current PDB?

HIV-1 protease structure is 1,157 as of 2019-05-07. Thousands of structural data from the same enzyme and virus. Shows a bit of redundancy.  See:http://www.rcsb.org/pdb/results/results.do?tabtoshow=Current&qrid=5C281786

**Task
Extract separate file for protein only

Extract Drug/Ligand only PDB

>Q6. How many amino acid residues are there in this pdb object and what are the two nonprotein residues?

Calpha indicates how many amino acids there are. The nonprotein residues are HOH(127) and MK1(1). 
```{r}
library(bio3d)

pdb <- read.pdb("1hsg.pdb")

attributes(pdb)

head(pdb$atom)

print(pdb)

```

```{r}
# Print a subset of $atom data for the first two atoms
pdb$atom[1:2, c("eleno", "elety", "x","y","z")]
```

```{r}
# Note that individual $atom records can also be accessed like this
pdb$atom$elety[1:2]
```


```{r}
# Which allows us to do the following
plot.bio3d(pdb$atom$b[pdb$calpha], sse=pdb, typ="l", ylab="B-factor")

```

>Q7. What type of R object is pdb$atom? HINT: You can always use the str() function to get a useful summery of any R object.

As you can see, it is a data.frame. 

```{r}
str(pdb$atom)
```

```{r}
# Print a summary of the coordinate data in $xyz
pdb$xyz

```

```{r}
# Examine the row and column dimensions
dim(pdb$xyz)

# Print coordinates for the first two atom
pdb$xyz[ 1, atom2xyz(1:2) ]
```

Atom selection is done via the function **atom.select()**
```{r}
# Select chain A
a.inds <- atom.select(pdb, chain="A")
# Select C-alphas of chain A
ca.inds <- atom.select(pdb, "calpha", chain="A")
# We can combine multiple selection criteria to return their intersection
cab.inds <- atom.select(pdb, elety=c("CA","CB"), chain="A",
resno=10:20)
```

The XYZ indices # is triple the atom indices. Three coordinates of XYZ for every one of these atoms. 
```{r}
#select protein only
prot.pdb <- atom.select(pdb, 'protein', value = TRUE)
write.pdb(prot.pdb, file = "1hsg_protein.pdb")
```

>Q8. Use the Bio3D write.pdb() function to write out a protein only PDB file for viewing in
VMD. Also write out a second separate PDB file for the ligand with residue name MK1
HINT: In Bio3D you can use the trim.pdb() function together with the pub object and your atom
selections such as atom.select(pdb, ’protein’) for the fist file and atom.select(pdb, ‘ligand’) for
the second.

```{r}
lig.pdb <- atom.select(pdb, "ligand", value = TRUE)
write.pdb(lig.pdb, file = "1hsg_ligand.pdb")
```



Section 5: Example Application on Adenylate Kinase 
```{r}
aa <- get.seq("1ake_A")

# Blast or hmmer search
b <- blast.pdb(aa)
```

```{r}
# Plot a summary of search results
hits <- plot(b)

```


```{r}
head(hits$pdb.id)

```

HWK: vmd - hsg section 1 and 2 

```{r}
# Fetch PDBs
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
# Align structures
pdbs <- pdbaln(files)
# Vector containing PDB codes
ids <- basename.pdb(pdbs$id)
# Draw schematic alignment
plot(pdbs, labels=ids)
```
```{r}
cons <- conserv(pdbs, method="entropy22")
# SSE annotations
sse <- pdbs2sse(pdbs, ind=1, rm.gaps=FALSE)
# Plot conservation per residue
plotb3(cons, sse=sse, ylab="Sequence entropy")
```

```{r}
anno <- pdb.annotate(ids)
print(unique(anno$source))

```

```{r}
library("bio3d.view")
pdb <- read.pdb("5p21")
view(pdb)
view(pdb,"overview", col="sse")
```


